### PR TITLE
Add input reader repository

### PR DESCRIPTION
**Pull request by GMKrieger**
#349

### CODE REVIEW

File: internal/repository/inputreader.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/cartesi/rollups-node/internal/node/model"
+	"github.com/jackc/pgx/v5"
+)
+
+func (pg *database) InsertInputsAndUpdateMostRecentFinalizedBlock(
```
Comment (fmoura) [inline internal/repository/inputreader.go:14]:
```suggestion
func (pg *database) InsertInputsAndUpdateMostRecentlyFinalizedBlock(
```

**Review by fmoura — CHANGES_REQUESTED** [review]

File: internal/repository/inputreader.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	ctx context.Context,
+	inputs []Input,
```
Comment (fmoura) [inline internal/repository/inputreader.go:16]:
```suggestion
	inputs []*Input,
```

Comment (fmoura) [reply internal/repository/inputreader.go:16]: Inputs have Blobs, thus can be large

**Review by fmoura — CHANGES_REQUESTED** [review]

File: internal/repository/inputreader.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	_, err = tx.Exec(ctx, query2, args)
+	if err != nil {
+		return fmt.Errorf("unable to insert inputs: %w\n", err)
+	}
+
+	err = tx.Commit(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to finish empty epoch: %w\n", err)
```
Comment (fmoura) [inline internal/repository/inputreader.go:67]:
```suggestion
		return fmt.Errorf("unable to insert inputs: %w\n", err)
```

**Review by fmoura — CHANGES_REQUESTED** [review]

File: internal/repository/inputreader_test.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+	var inputs []Input
+	inputs = append(inputs, input1)
+	inputs = append(inputs, input2)
+	inputs = append(inputs, input3)
+
+	err := s.database.InsertInputsAndUpdateMostRecentFinalizedBlock(s.ctx, inputs, 5)
+	s.Require().Nil(err)
+
+	response, err := s.database.GetInput(s.ctx, 6)
```
Comment (fmoura) [inline internal/repository/inputreader_test.go:44]:
```suggestion
	response, err := s.database.GetInput(s.ctx, 6)
	s.Require().Nil(err)
```

File: internal/repository/inputreader_test.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+func (s *RepositorySuite) TestInsertInputsAndUpdateMostRecentFinalizedBlockEmptyInputs() {
+	var inputs []Input
+
+	err := s.database.InsertInputsAndUpdateMostRecentFinalizedBlock(s.ctx, inputs, 6)
+	s.Require().Nil(err)
+
+	var block uint64 = 6
+	response, err := s.database.GetMostRecentBlock(s.ctx)
```
Comment (fmoura) [inline internal/repository/inputreader_test.go:55]:
```suggestion
	response, err := s.database.GetMostRecentBlock(s.ctx)
	s.Require().Nil(err)
```

**Review by fmoura — CHANGES_REQUESTED** [review]

**Comment by fmoura** [comment]
When running the `cartesi-rollups-node` for a second time using the same db instance , This error occurs :
```
2024-06-25T23:07:57.250 ERR Unable to run migrations error="no change"
```
This error should be treated differently, as it will happen on a node restart

**Review by fmoura — DISMISSED** [review]

File: internal/repository/inputreader.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	"context"
+	"fmt"
+
+	. "github.com/cartesi/rollups-node/internal/node/model"
+	"github.com/jackc/pgx/v5"
+)
+
+func (pg *database) InsertInputsAndUpdateMostRecentlyFinalizedBlock(
+	ctx context.Context,
+	inputs []*Input,
```
[the remark this answers is not in the document]
Comment (fmoura) [reply internal/repository/inputreader.go:16]: Here the pointers are more about not copying big structures than using pointer as an "optional" input. Inputs have a `blob []byte` inside.

File: internal/repository/inputreader.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	args := pgx.NamedArgs{
+		"blockNumber": blockNumber,
+	}
+
+	tx, err := pg.db.Begin(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to insert inputs: %w\n", err)
+	}
+
+	if len(inputs) != 0 {
```
[the remark this answers is not in the document]
Comment (fmoura) [reply internal/repository/inputreader.go:44]: If you mean the Input part, I agree. But the most recently finalized block can be updated without adding new inputs

File: internal/repository/inputreader_test.go
[the remark this answers is not in the document]
Comment (GMKrieger) [reply internal/repository/inputreader_test.go:1]: The 2 tests that check the most recent finalized block need checks on the database to work. I've created an issue to keep track of this for a future PR. #497 

For now, I will create the first 2 tests for this PR.

**Comment by fmoura** [comment]
When trying to Setup the database with "actual" values, like `deployment_block = 0x10` I am getting a fk error : 
```
"unable to setup database state: ERROR: insert or update on table \"node_state\" violates foreign key constraint \"node_state_current_epoch_fkey\" (SQLSTATE 23503)\n"
```

I seems it is expecting all epochs from block zero til deployment block to be at the database

File: internal/repository/inputreader.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (torives) [reply internal/repository/inputreader.go:16]: The new slice has a reference to the same backing array, so there are no blobs being copied here 😉. [Source](https://go.dev/blog/slices-intro)

**Comment by GMKrieger** [comment]
> When trying to Setup the database with "actual" values, like `deployment_block = 0x10` I am getting a fk error :
> 
> ```
> "unable to setup database state: ERROR: insert or update on table \"node_state\" violates foreign key constraint \"node_state_current_epoch_fkey\" (SQLSTATE 23503)\n"
> ```
> 
> I seems it is expecting all epochs from block zero til deployment block to be at the database

This error is due to the foreign key of the epoch not being followed. When setting up the database, you have to be sure the first block of the epoch and the current epoch (which on the setup is the deployment block) is the same.

This will change for the next update of the database for the multi-dapp.

File: internal/repository/inputreader.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (fmoura) [reply internal/repository/inputreader.go:16]: The Blob itself is a slice. So my main concern do not apply . Thanks

**Comment by fmoura** [comment]
Got another error while testing. 

Every time the node is started, the Repository Setup function is always adding a new config row, even if it is a node restart ( when the same Postgres instance is the same ) . It should be checked and then just logged

**Comment by fmoura** [comment]
Also just a comment: 

Using go test coverage during tests, I checked that the tests are not exercising , at the `InsetOutput` method the control flow when a output is not `verifiable`

File: internal/repository/inputreader.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (fmoura) [reply internal/repository/inputreader.go:16]: @GMKrieger this suggestion can be accepted. There is no need to use pointers

**Comment by GMKrieger** [comment]
This PR is deprecated with the release of the multi-dapp repository #511.

**Pull request closed without merging**